File: backend/domains/gateway/domain/thinking_param.py
"""思考模式参数类型推断（与 GatewayModel.tags / 调用指南、Playground 对齐）。"""

from __future__ import annotations

from typing import Any

from domains.gateway.domain.temperature_policy import enrich_temperature_tags

THINKING_PARAM_NONE = "none"
THINKING_PARAM_DASHSCOPE = "dashscope_enable_thinking"
THINKING_PARAM_BUILTIN = "builtin_reasoning"
THINKING_PARAM_ANTHROPIC = "anthropic_extended"
THINKING_PARAM_DEEPSEEK_V4 = "deepseek_v4_thinking"

# 运营显式写入 thinking_param=none 时置 true，阻止按 real_model 重新推断（GatewayModel.tags）。
THINKING_PARAM_LOCKED_TAG = "thinking_param_locked"

THINKING_PARAM_VALUES: frozenset[str] = frozenset(
    {
        THINKING_PARAM_NONE,
        THINKING_PARAM_DASHSCOPE,
        THINKING_PARAM_BUILTIN,
        THINKING_PARAM_ANTHROPIC,
        THINKING_PARAM_DEEPSEEK_V4,
    }
)


def is_deepseek_v4_model_id(model_id: str) -> bool:
    """DeepSeek V4 Pro/Flash：OpenAI 兼容入口 + ``extra_body.thinking`` 可开关。

    前端 ``model-selector-capabilities`` 的弱推断须与此规则保持同步。
    """
    lower = (model_id or "").strip().lower()
    if not lower:
        return False
    model_part = lower.split("/")[-1]
    return "deepseek-v4-pro" in model_part or "deepseek-v4-flash" in model_part


def _model_id_lower(provider: str, real_model: str) -> str:
    rm = (real_model or "").strip().lower()
    if "/" in rm:
        return rm
    p = (provider or "").strip().lower()
    return f"{p}/{rm}" if p and rm else rm


def _normalized_thinking_param(raw: object) -> str | None:
    if not isinstance(raw, str):
        return None
    normalized = raw.strip()
    return normalized if normalized in THINKING_PARAM_VALUES else None


def _thinking_param_locked(tags: dict[str, Any]) -> bool:
    return tags.get(THINKING_PARAM_LOCKED_TAG) is True


def infer_thinking_param(
    *,
    provider: str,
    real_model: str,
    supports_reasoning: bool = False,
    explicit: str | None = None,
) -> str:
    """按 provider / real_model 推断思考模式；``explicit`` 为 tags 中显式配置时优先。"""
    if explicit is not None:
        normalized = explicit.strip()
        if normalized in THINKING_PARAM_VALUES:
            return normalized

    combined = _model_id_lower(provider, real_model)
    model_part = combined.split("/")[-1] if combined else ""
    prov = (provider or "").strip().lower()

    if prov == "dashscope":
        if "qwq" in model_part:
            return THINKING_PARAM_BUILTIN
        if "qwen3" in model_part or model_part.startswith("qwen-3"):
            return THINKING_PARAM_DASHSCOPE

    if is_deepseek_v4_model_id(combined) or is_deepseek_v4_model_id(model_part):
        return THINKING_PARAM_DEEPSEEK_V4

    if prov == "deepseek" and "reasoner" in combined:
        return THINKING_PARAM_BUILTIN

    if prov == "anthropic" and supports_reasoning:
        return THINKING_PARAM_ANTHROPIC

    if supports_reasoning:
        return THINKING_PARAM_BUILTIN

    return THINKING_PARAM_NONE


def resolve_thinking_param_from_tags(
    tags: dict[str, Any],
    *,
    provider: str = "",
    real_model: str = "",
) -> str:
    """从 tags 解析 thinking_param；读写路径 SSOT（enrich 与出站快照共用）。"""
    explicit = _normalized_thinking_param(tags.get("thinking_param"))
    supports_reasoning = bool(tags.get("supports_reasoning", False))
    if bool(tags.get("supports_reasoning_content", False)):
        supports_reasoning = True

    prov = str(tags.get("provider") or provider or "")
    rm = str(tags.get("real_model") or real_model or "")
    inferred = infer_thinking_param(
        provider=prov,
        real_model=rm,
        supports_reasoning=supports_reasoning,
        explicit=None,
    )

    if explicit is not None and explicit != THINKING_PARAM_NONE:
        return explicit
    if explicit == THINKING_PARAM_NONE and _thinking_param_locked(tags):
        return THINKING_PARAM_NONE
    return inferred


def enrich_gateway_model_tags(
    tags: dict[str, Any] | None,
    *,
    provider: str,
    real_model: str,
) -> dict[str, Any]:
    """注册/更新模型时合并 ``thinking_param`` 与 ``supports_reasoning``（权威写入 tags）。"""
    merged: dict[str, Any] = dict(tags or {})
    merged["provider"] = provider
    merged["real_model"] = real_model

    thinking_param = resolve_thinking_param_from_tags(
        merged, provider=provider, real_model=real_model
    )
    merged["thinking_param"] = thinking_param
    merged["supports_reasoning"] = effective_supports_reasoning(merged, thinking_param)
    return enrich_temperature_tags(merged, thinking_param=thinking_param)


def effective_supports_reasoning(tags: dict[str, Any], thinking_param: str) -> bool:
    """与 thinking_param 对齐的 supports_reasoning（用于目录同步）。"""
    if bool(tags.get("supports_reasoning", False)):
        return True
    if bool(tags.get("supports_reasoning_content", False)):
        return True
    return thinking_param != THINKING_PARAM_NONE


__all__ = [
    "THINKING_PARAM_ANTHROPIC",
    "THINKING_PARAM_BUILTIN",
    "THINKING_PARAM_DASHSCOPE",
    "THINKING_PARAM_DEEPSEEK_V4",
    "THINKING_PARAM_LOCKED_TAG",
    "THINKING_PARAM_NONE",
    "THINKING_PARAM_VALUES",
    "effective_supports_reasoning",
    "enrich_gateway_model_tags",
    "infer_thinking_param",
    "is_deepseek_v4_model_id",
    "resolve_thinking_param_from_tags",
]

File: frontend/src/features/gateway-shared/thinking-param.ts
/**
 * 思考模式 UI 文案与 Playground 提示（能力解析见 model-selector-capabilities）。
 */

import {
  resolveThinkingParamForModel,
  thinkingParamFromCapabilities,
  type ThinkingParam,
} from './model-selector-capabilities'

export type { ThinkingParam } from './model-selector-capabilities'
export {
  resolveThinkingParamForModel,
  thinkingParamFromCapabilities,
} from './model-selector-capabilities'

/** 模型列表芯片展示文案 */
export function thinkingParamLabel(param: ThinkingParam): string | null {
  switch (param) {
    case 'dashscope_enable_thinking':
      return 'Qwen3 思考'
    case 'builtin_reasoning':
      return '内置推理'
    case 'anthropic_extended':
      return 'Extended Thinking'
    case 'deepseek_v4_thinking':
      return 'V4 思考'
    default:
      return null
  }
}

export function thinkingHintForModel(
  modelName: string,
  apiFlavor: 'openai' | 'anthropic',
  capabilities?: Record<string, unknown>,
  options?: { allowNameFallback?: boolean }
): string | null {
  const param = resolveThinkingParamForModel(modelName, capabilities, options)
  const fromApi = thinkingParamFromCapabilities(capabilities) !== null

  if (param === 'none') {
    return `当前模型「${modelName}」通常不支持思考模式；请勿对普通模型传 enable_thinking / thinking。`
  }
  if (apiFlavor === 'anthropic') {
    if (param === 'anthropic_extended') {
      return `当前模型「${modelName}」：推荐在请求体使用 thinking: { type: "enabled", budget_tokens: 8000 }。`
    }
    if (param === 'deepseek_v4_thinking') {
      return `当前模型「${modelName}」为 DeepSeek V4，请使用 OpenAI 兼容入口并传 extra_body.thinking（Anthropic 入口不适用）。`
    }
    return `当前模型「${modelName}」：Anthropic 原生 thinking 通常用于 Claude；请确认模型能力。`
  }
  switch (param) {
    case 'dashscope_enable_thinking':
      return `当前模型「${modelName}」：推荐 stream: true + enable_thinking / extra_body.enable_thinking: true；非流式须 enable_thinking: false。${fromApi ? '' : '（能力来自模型名推断，请以模型列表标签为准）'}`
    case 'builtin_reasoning':
      return `当前模型「${modelName}」：内置推理，无需 enable_thinking；流式/非流式均可读 reasoning_content。`
    case 'deepseek_v4_thinking':
      return `当前模型「${modelName}」：推荐 extra_body.thinking: { type: "enabled" }（可选 reasoning_effort: "high"）；响应含 reasoning_content。${fromApi ? '' : '（能力来自模型名推断，请以模型列表标签为准）'}`
    default:
      return null
  }
}

/** Playground：内置推理模型说明（无请求开关） */
export function builtinReasoningPlaygroundHint(param: ThinkingParam): string | null {
  if (param !== 'builtin_reasoning') return null
  return '内置推理模型：无需开启思考模式开关，响应中的 reasoning_content 会自动展示。'
}
